kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+        if result == -1:
+            logger.error(
+                'Trial did not converge:\n'
+                f'    `train_info.best_metrics={train_info.best_metrics}`\n'
+                f'Setting result to -1'
+            )
+
+        logger.info(f'RESULTS: {result} {self.main_metric}')
+        logger.info(f'MODEL: {train_info.best_model_path}')
+        logger.info(f'PARAMETERS: {search_values}')
+        for name, value in search_values.items():
+            logger.info(f'{name}: {value}')
+
+        # Store the checkpoint path for later access
+        self.model_paths.append((train_info.best_model_path, result))
+
+        return result
+
+
+def setup_run(directory: Path, seed: int, debug=False, quiet=False) -> Path:
+    # TODO: replace all of this with the MLFlow integration: optuna.integration.mlflow
+    #   In particular: let the output directory be created by MLflow entirely
+    if not directory.exists():
+        # Initialize a new folder with name '0'
+        output_dir = directory / '0'
+        output_dir.mkdir(parents=True)
+    else:
+        # Initialize a new folder incrementing folder count
+        output_dir = directory / str(len(list(directory.glob('*'))))
+        if output_dir.exists():
+            backup_directory = output_dir.with_name(
+                f'{output_dir.name}_backup_{datetime.now().isoformat()}'
+            )
+            logger.warning(
+                f'Folder {output_dir} already exists; moving it to {backup_directory}'
+            )
+            output_dir.rename(backup_directory)
+        output_dir.mkdir(parents=True)
+
+    logger.info(f'Initializing new search folder at: {output_dir}')
+
+    configure_logging(output_dir=output_dir, verbose=debug, quiet=quiet)
+    configure_seed(seed)
+
+    return output_dir
+
+
+def run(config: Configuration):
+    try:
+        import optuna
+    except ImportError as e:
+        raise ImportError(
+            f'{e}. Install the search dependencies with '
+            '``pip install -U openkiwi[search]``, or with '
+            '``poetry install -E search`` when setting '
+            'up for local development'
+        )
+
+    output_dir = setup_run(config.directory, config.seed)
+    logger.info(f'Saving all Optuna results to: {output_dir}')
+
+    if isinstance(config.base_config, Path):
+        # FIXME: this is not neat (but it does work)
+        #   We need this in order to support the `defaults` field in the train config,
+        #   like the
+        #     defaults:
+        #        - data: wmt19.qe.en_de
+        #   in the config/xlmroberta. It's Hydra that takes care of that inside
+        #   `kiwi.cli.arguments_to_configuration`. And because Hydra has already
+        #   been configured globally in the Kiwi cli, we need to clear it.
+        hydra._internal.hydra.GlobalHydra().clear()
+        base_dict = kiwi.cli.arguments_to_configuration(
+            {'CONFIG_FILE': config.base_config}
+        )
```

Comment:
Change this to `kiwi.lib.utils.file_to_configuration`

Reply:
Will wait till that branch is merged